zkvm/src/types.rs

```diff
@@ -136,6 +139,43 @@ impl ScalarWitness {
     }
 }
 
+impl Neg for ScalarWitness {
+    type Output = ScalarWitness;
+
+    fn neg(self) -> ScalarWitness {
+        match self {
+            ScalarWitness::Integer(a) => ScalarWitness::Integer(-a),
+            ScalarWitness::Scalar(a) => ScalarWitness::Scalar(-a),
+        }
+    }
+}
+impl Add for ScalarWitness {
+    type Output = ScalarWitness;
+
+    fn add(self, rhs: ScalarWitness) -> ScalarWitness {
+        match (self, rhs) {
+            (ScalarWitness::Integer(a), ScalarWitness::Integer(b)) => {
+                let res = a + b;
+
+                if res > spacesuit::SignedInteger::from(std::u64::MAX) {
+                    return ScalarWitness::Scalar(res.into());
+                }
+
+                ScalarWitness::Integer(res)
+            }
+            (ScalarWitness::Scalar(a), ScalarWitness::Scalar(b)) => ScalarWitness::Scalar(a + b),
+            (ScalarWitness::Integer(a), ScalarWitness::Scalar(b)) => {
+                let x: Scalar = a.into();
+                ScalarWitness::Scalar(x + b)
```

Comment:
let's keep a one-liner (similarly for another branch below):

```rust
 (ScalarWitness::Integer(a), ScalarWitness::Scalar(b)) => ScalarWitness::Scalar(Scalar::from(a) + b)
```

Reply:
I'm not sure how to do it with a one liner. If I understand correctly `impl Into<Scalar> for SignedInteger` would not give me the `From` trait for free, but an `impl From` will give me the `Into` trait`?

Reply:
Oh, you can also try `Into<Scalar>::into(a)` if `a.into()` is ambiguous to the compiler. 

Cf. [Fully Qualified Syntax for Disambiguation](https://doc.rust-lang.org/1.22.1/book/second-edition/ch19-03-advanced-traits.html#fully-qualified-syntax-for-disambiguation)

Reply:
In fact, we can probably do 2 arms in this match statement: one is for `Integer,Integer`, and the other is for all other combinations: `(a,b) => ScalarWitness::Scalar(a.into() + b.into())`.